homeassistant/components/dsmr/sensor.py

```diff
@@ -431,41 +431,42 @@ def rename_old_gas_to_mbus(
 ) -> None:
     """Rename old gas sensor to mbus variant."""
     dev_reg = dr.async_get(hass)
-    device_entry_v1 = dev_reg.async_get_device(identifiers={(DOMAIN, entry.entry_id)})
-    if device_entry_v1 is not None:
-        device_id = device_entry_v1.id
-
-        ent_reg = er.async_get(hass)
-        entries = er.async_entries_for_device(ent_reg, device_id)
-
-        for entity in entries:
-            if entity.unique_id.endswith(
-                "belgium_5min_gas_meter_reading"
-            ) or entity.unique_id.endswith("hourly_gas_meter_reading"):
-                try:
-                    ent_reg.async_update_entity(
-                        entity.entity_id,
-                        new_unique_id=mbus_device_id,
-                        device_id=mbus_device_id,
-                    )
-                except ValueError:
-                    LOGGER.debug(
-                        "Skip migration of %s because it already exists",
-                        entity.entity_id,
-                    )
-                else:
-                    LOGGER.debug(
-                        "Migrated entity %s from unique id %s to %s",
-                        entity.entity_id,
-                        entity.unique_id,
-                        mbus_device_id,
-                    )
-        # Cleanup old device
-        dev_entities = er.async_entries_for_device(
-            ent_reg, device_id, include_disabled_entities=True
-        )
-        if not dev_entities:
-            dev_reg.async_remove_device(device_id)
+    for dev_id in (mbus_device_id, entry.entry_id):
```

Comment:
We could look for a device with entry.data[CONF_SERIAL_ID_GAS]?

Reply:
In fact we should stop using CONF_SERIAL_ID_GAS for devices with mbus devices.
Cause `CONF_SERIAL_ID_GAS` is a single entity ... while with mbus we have 4 devices that can be linked to the main device, so you can have 4 gas meters in theory.
Better check for the mbus_device_id.

Reply:
What about non mbus devices.

Reply:
How you mean? There should be only 1 main SERIAL for the device and then a serial per mbus device.

Reply:
I am not sure what il saying. I think i would have been more happy with doing a full unique_id check and just iterate over all entities in the config entry instead of per device.

Unique_id, should be just that,  unique for the integration. The associated device have nothing to do with it really.

Reply:
The replacements should be static once we know the serial.

{serial_gas}_hourly.. -> serial_gas.
{serial_main_hourly} -> ? (Do we even know these exist?)

Reply:
But there is no advantage afaik to loop over all entities. Cause with this patch we loop over all the devices that can match already.
If there would be more dsmr devices, this won't match anyway when doing a full unique_id check, as the serial in it should be the same as the device serial, and thus different anyway. So useless to lookup.

Only since https://github.com/home-assistant/core/commit/91182603d5756c38714f4c21d5cd6e04f1522d2b the correct serial is used for gas meters in 5B. That's why the migration code was working fine for 5B in the past, as entry_id was used and not the correct serial.
But for 5 (and now also 5B), the real gas serial is used, so we need to use that to lookup the device.

Also note that with supporting multiple mbus devices, its possible that you now have 1 gas meter, but with the multiple mbus support, it will see both gas meters.
So we need to make sure here to migrate the existing sensor to the correct device.

Reply:
What was thinking was this. The update of which device they belong to does not to be done since that will be handled once they get added to the system. I find this much more readable. (not fully fleshed out the code, so tests fail due to not adjusting the associated device id)

The problem here is that: "{entry.entry_id}_hourly_gas_meter_reading" can exist in 4 series device without any mbus data: See here: https://github.com/ndokter/dsmr_parser/blob/94189b6b63e12554c866bc2964ffb7d5c7e8b2c4/dsmr_parser/telegram_specifications.py#L286

So i'm not sure we should include that...

```python
def rename_old_gas_to_mbus(
    hass: HomeAssistant, entry: ConfigEntry, mbus_device_id: str
) -> None:
    """Rename old gas sensor to mbus variant."""

    adjustments = {
        f"{mbus_device_id}_hourly_gas_meter_reading": mbus_device_id,
        f"{entry.entry_id}_hourly_gas_meter_reading": mbus_device_id,
        f"{entry.entry_id}_belgium_5min_gas_meter_reading": mbus_device_id,
        f"{mbus_device_id}_belgium_5min_gas_meter_reading": mbus_device_id,
    }

    ent_reg = er.async_get(hass)
    entries = er.async_entries_for_config_entry(ent_reg, entry.entry_id)
    for entity in entries:
        if entry.unique_id and (new := adjustments.get(entry.unique_id, None)):
            try:
                ent_reg.async_update_entity(
                    entity.entity_id,
                    new_unique_id=new,
                )
            except ValueError:
                LOGGER.debug(
                    "Skip migration of %s because it already exists",
                    entity.entity_id,
                )
            else:
                LOGGER.debug(
                    "Migrated entity %s from unique id %s to %s",
                    entity.entity_id,
                    entity.unique_id,
                    mbus_device_id,
                )
```

Reply:
But migrating the sensor to the correct/new device is better/cleaner, as before https://github.com/home-assistant/core/commit/91182603d5756c38714f4c21d5cd6e04f1522d2b for example the device did not have the correct serial like it should.
The current migration code takes care of that also.

If there is no mbus data, then the migration code will not be run. So that shouldn't cause an issue?

Reply:
I still don't like doing a `RANDOM_PREFIX_THAT_CAN_BE_ANYTHINGhourly_gas_meter_reading` -> `mbus_device_id`. It's very error prone. 

ALSO, which missed before: async_entries_for_device can include entities from other platform, like helpers. They are unlikely to have that unique_id, but nothing really stopping them.

Reply:
> But migrating the sensor to the correct/new device is better/cleaner, as before https://github.com/home-assistant/core/commit/91182603d5756c38714f4c21d5cd6e04f1522d2b for example the device did not have the correct serial like it should.
This will be handled automatically as soon as the device is added to the system so migration code does not need to do that.

Reply:
> > But migrating the sensor to the correct/new device is better/cleaner, as before [9118260](https://github.com/home-assistant/core/commit/91182603d5756c38714f4c21d5cd6e04f1522d2b) for example the device did not have the correct serial like it should.
> 
> This will be handled automatically as soon as the device is added to the system so migration code does not need to do that.

As long as we cleanup the old/empty device then :)

Reply:
> I still don't like doing a `RANDOM_PREFIX_THAT_CAN_BE_ANYTHINGhourly_gas_meter_reading` -> `mbus_device_id`. It's very error prone.
> 
> ALSO, which missed before: async_entries_for_device can include entities from other platform, like helpers. They are unlikely to have that unique_id, but nothing really stopping them.

Just need to make sure there are not older versions that created the sensors with even another prefix.
I think not, but might be good to double check.

Reply:
Ok. Let's do it this way for now. The prefix was weird historically too.

I would like to remove this migration code as soon as we can. But ends up a year in the future i think.

@dupondje  Think you would be up for taking on maintainership on this component?